crates/sqlbuiltins/src/functions/table/object_store.rs

```diff
@@ -582,3 +585,122 @@ fn create_azure_store_access(
+        Ok(RuntimePreference::Remote)
+    }
+
+    async fn create_provider(
+        &self,
+        ctx: &dyn TableFuncContextProvider,
+        args: Vec<FuncParamValue>,
+        _opts: HashMap<String, FuncParamValue>,
+    ) -> Result<Arc<dyn TableProvider>> {
+        if args.len() != 1 {
+            return Err(ExtensionError::InvalidNumArgs);
+        }
+
+        let session_state = ctx.get_session_state();
+
+        // NativeTableStorage is available on Remote ctx. Use store already
+        // constructed there.
+        let storage = ctx
+            .get_session_state()
+            .config()
+            .get_extension::<NativeTableStorage>()
+            .ok_or_else(|| {
+                ExtensionError::String(
+                    format!(
+                        "access unavailable, {} is not supported in local environments",
+                        self.name(),
+                    )
+                    .to_string(),
+                )
+            })?;
+        let store = storage.store.clone(); // Cheap to clone
+
+        let file_name: String = args.into_iter().next().unwrap().try_into()?;
+        let ext = Path::new(&file_name)
+            .extension()
+            .ok_or_else(|| {
+                ExtensionError::String(
+                    "missing file extension, supported: [.csv, .json, .parquet]".to_string(),
+                )
+            })?
+            .to_str()
+            .ok_or_else(|| {
+                ExtensionError::String(format!("unsupported file extension: {file_name}"))
+            })?
+            .to_lowercase();
+        let file_format: Arc<dyn FileFormat> = match ext.as_str() {
+            "csv" => Arc::new(CsvFormat::default().with_schema_infer_max_rec(Some(20480))),
+            "json" => Arc::new(JsonFormat::default()),
+            "parquet" => Arc::new(ParquetFormat::default()),
+            ext => {
+                return Err(ExtensionError::String(format!(
+                    "unsupported file extension: {ext}"
+                )))
+            }
+        };
+
+        // This is required to read the object meta, but we unfortunately also
+        // need the base_url below for ObjStoreTableProvider impl. Maybe there's
+        // a refactor opportunity.
+        let prefix: ObjectStorePath =
+            format!("databases/{}/uploads/{}", storage.db_id(), file_name).into();
+        let base_url = format!("{}", storage.root_url);
+        let base_url = ObjectStoreUrl::parse(base_url)?;
+
+        // glaredb_upload currently does not support globbing, and therefore we
+        // do not need to iterate: there should only be one.
+        let mut objects = store.list(Some(&prefix));
+        let meta = objects
+            .next()
+            .await
+            .ok_or_else(|| ExtensionError::String(format!("file not found: {}", file_name)))?
+            .or_else(|_| return Err(ExtensionError::ObjectStore("read file".to_string())))?;
```

Comment:
this will fail clippy, I believe. Need to figure out canonical way to handle an `Option<Result>`